### PR TITLE
feat: special case arb search

### PR DESCRIPTION
**Pull request by cartcrom**
## Description

Special cases the ARB token on Arbitrum to always display over the ARB token on Mainnet in search results when both are returned by the backend, regardless of the user's current chain.

Also cleans up / renames the cross token search code to make functionality more clear.

_[Linear ticket](https://linear.app/uniswap/issue/WEB-1949/special-case-dollararb-in-search-so-the-arbitrum-network-one-shows-up)_

## Screen capture

| Before       | After (Desktop) |
| ------------ |---------------- |
| image |  image      |

### QA (ie manual testing)

- [ ] Search for a string that should return ARB as a search result, ensure it has the Arbitrum logo rather than no logo

### CODE REVIEW

**Review by cbachmeier — COMMENTED** [review]
I'd like to avoid 1 offs whenever possible. MATIC is cross chain, what's the logic we use for this token to show it's native network first and can we re-use it for ARB?
Screen Shot 2023-05-16 at 15 07 11

**Comment by cartcrom** [comment]
> I'd like to avoid 1 offs whenever possible. MATIC is cross chain, what's the logic we use for this token to show it's native network first and can we re-use it for ARB? Screen Shot 2023-05-16 at 15 07 11

I completely agree about 1 offs, and suggested using huge volume differences as a metric to override the default behaviour of showing the token on the wallet's current chain, but product wants to keep the scope of this change to just ARB for now. We can't re-use the MATIC logic, because that logic relies on Polygon MATIC being a native token

**Review by cbachmeier — CHANGES_REQUESTED** [review]

**Review by cbachmeier — APPROVED** [review]
Returns expected result, however not sure if second check is necessary

File: src/graphql/data/SearchTokens.ts [hunk 1]
```diff
@@ -41,19 +42,30 @@ gql`
...
-function isMoreRevelantToken(current: SearchToken, existing: SearchToken | undefined, searchChain: Chain) {
-  if (!existing) return true
+/* Returns the more relevant cross-chain token based on native status and search chain */
+function dedupeCrosschainTokens(current: SearchToken, existing: SearchToken | undefined, searchChain: Chain) {
+  if (!existing) return current
+  invariant(current.project?.id === existing.project?.id, 'Cannot dedupe tokens within different tokenProjects')
+
+  // Special case: always prefer Arbitrum ARB over Mainnet ARB
+  if (current.address?.toLowerCase() === ARB_ADDRESS) return current
+  if (existing.address?.toLowerCase() === ARB_ADDRESS) return existing
```
Comment (cbachmeier) [inline src/graphql/data/SearchTokens.ts:56]: Are you sure both of these are necessary? Having just the first check, `if (current.address?.toLowerCase() === ARB_ADDRESS) return current`, with the lowercase casting works for mainnet and arbitrum networks
Screen Shot 2023-05-18 at 13 42 25

Comment (cartcrom) [reply src/graphql/data/SearchTokens.ts:56]: It depends on the order the backend returns the two ARB tokens in, if the backend started returning them in the opposite order, only having one check wouldn't work.

File: src/graphql/data/SearchTokens.ts [hunk 2]
```diff
@@ -87,7 +99,7 @@ export function useSearchTokens(searchQuery: string, chainId: number) {
     data?.searchTokens?.forEach((token) => {
       if (token.project?.id) {
         const existing = selectionMap[token.project.id]
-        if (isMoreRevelantToken(token, existing, searchChain)) selectionMap[token.project.id] = token
```
Comment (JFrankfurt) [inline src/graphql/data/SearchTokens.ts:90]: it looks like there's a behavior change here that might not be intentional--in the old one there were cases where `selectionMap[token.project.id]` wouldn't be assigned a value here, but now every iteration assigns a value. Is that intended?

Comment (cartcrom) [reply src/graphql/data/SearchTokens.ts:90]: In cases where `isMoreRevelantToken` would've returned false, `dedupeCrosschainTokens` returns `existing`, so this is just a refactor for readability

**Review by JFrankfurt — APPROVED** [review]

**Comment by JFrankfurt** [comment]
smol

**Pull request merged**